src/tds/convert.c

```diff
@@ -78,26 +78,37 @@ static const char *parse_numeric(const char *buf, const char *pend,
 
 #define test_alloc(x) {if ((x)==NULL) return TDS_CONVERT_NOMEM;}
 
-#define IS_TINYINT(x) ( 0 <= (x) && (x) <= 0xff )
-#define IS_SMALLINT(x) ( -32768 <= (x) && (x) <= 32767 )
-#define IS_USMALLINT(x) ( 0 <= (x) && (x) <= 65535 )
+/*
+ * Confirm convertability to a signed type with the given bounds,
+ * bearing in mind that promotion rules can complicate comparison
+ * between negative and unsigned integers, that conversion from
+ * floating-point values to integers starts by truncating fractional
+ * portions, and that lb - 1 might have the same FP representation as
+ * lb in x's type (as with float -> int and anything -> TDS_INT8).
+ */
+#define IS_INTn(x, lb, ub) ((x) < 0 ? lb - (x) < 1 : (x) < ub + 1)
```

Comment:
To be honest this looks complicated... and also I don't think compilers can reach the level of the 2 initial simple comparisons. And the `lb - (x)` part could be subject to loose of precision.

Reply:
The general idea with this macro is to unify cases that would otherwise require slightly different pairs of comparisons.  In particular:
- For `int`, `(x) >= TDS_INT_MIN` is correct for every source type but `double`, where it misses the open interval `(TDS_INT_MIN - 1, TDS_INT_MIN)`, and `(x) > (TDS_INT8) TDS_INT_MIN - 1` is correct for every source type but `float`, where `TDS_INT_MIN - 1` is unrepresentable and typically rounds to `TDS_INT_MIN`.  Looking at `lb - (x)` should yield correct results in either case, but will misfire if `x` is `unsigned int` or `TDS_UINT8`, necessitating upfront comparison with zero to be fully general.
- For `TDS_INT8`, `(x) <= TDS_INT8_MAX` picks up floating-point values that are slightly out of range and will wrap around upon conversion; `(x) < (TDS_UINT8) TDS_INT8_MAX + 1` fixes that but proceeds to reject negative integer values, likewise necessitating upfront comparison with zero to be fully general.

At any rate, I'm comparing `lb - (x) < 1` rather than `lb - 1 < (x)` due to precision concerns -- if `lb - 1` loses precision, the comparison could (and typically would) wrongly exclude `lb`, whereas `lb - (x)` will lose precision only if the difference is very positive (and hence clearly greater than one) or very negative (and hence clearly less than one).

Reply:
I tried something similar under compiler explorer:
```c++
#define INSIDE(x) ((x) > -4 && (x) < 10)

bool limit1(int a) {
    return INSIDE(a);
}

bool limit2(double a) {
    return INSIDE(a);
}

#define IS_INTn(x, lb, ub) ((x) < 0 ? lb - (x) < 1 : (x) < ub + 1)
#define INSIDE2(a) IS_INTn(a, -4, 10)

bool limit3(int a) {
    return INSIDE2(a);
}

bool limit4(double a) {
    return INSIDE2(a);
}
```
and in all compilers I tried the code is less efficient than normal bound checks (as I expected).
I wrote this https://github.com/freddy77/freetds/commit/7ebc5d85e6c001838bd2f6cba4bd623f5e5eea48 to simplify and use simpler bound checks. Please consider to squash it in.

Reply:
Squashed, thanks, with some fine-tuning to account for pre-conversion truncation of negative FP values (which I'd previously missed myself for unsigned destination types).